### PR TITLE
fix: invokeAndWait2 should throw the exception in the tasks

### PR DESCRIPTION
**Pull request by i8run**
## What changes were proposed in this pull request?

The `invokeAndWait2` will not throw exceptions in the tasks to caller thread.

## How was this patch tested?
test cases and Jenkins.

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 1]
```diff
@@ -556,7 +556,24 @@ object Engine {
...
+    val affinityCores = Affinity.getAffinity
+    val physicalCoreNum = CpuInfo.getPhysicalProcessorCount
+    val affinityCoreNum = affinityCores.length
+
+    // 1. this library in docker/cgroup env, which sets cpu affinity fist. so we can't use
+    //    resources exceeding limits.
+    // 2. this library is in a hyper thread envs, which enables hyper threading of cpu. so
+    //    we should set the mkl num threads to physical core number for performance
+
+    val default = if (affinityCores.min > 0 && affinityCores.max >= physicalCoreNumber) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: When it happens `affinityCores.max >= physicalCoreNumber`

Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: hyper threading is enabled.

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: ok, why not use physicalCoreNumber to bind affinity in this case.

Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: this case can't use. for example there're 8 logical cores and 4 physical cores. If the affinity settings is 2-6 (starting from 0), which means this environment only can use core 2, 3, 4. So I can't set the `physicalCoreNumber`.

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: ok， I see

**Pull request merged**